src/lib/fountain/fountain_decoder_sink.h

```diff
@@ -23,13 +24,35 @@ std::function<void(const std::string&, const std::vector<uint8_t>&)> write_on_st
 		f.write((char*)data.data(), data.size());
 		if (log_writes)
 			printf("%s\n", file_path.c_str());
+		return filename;
 	};
 }
 
+template <typename OUTSTREAM>
+std::function<std::string(const std::string&, const std::vector<uint8_t>&)> decompress_on_store(std::string data_dir, bool log_writes=false)
+{
+	return [data_dir, log_writes](const std::string& fallback_name, const std::vector<uint8_t>& data)
+	{
+		std::string filename = cimbar::zstd_header_check::get_filename(data.data(), data.size());
```

Comment:
We could attack this a couple ways:
1. when decompressing, read everything once, and check the opening read for our metadata. This would require lazy opening the file when writing. (e.g. we'd have to open on write(), not open())
2. read the opening bytes twice. Once here, and again in the decompressor.

The key insight here (imo) is that the input is already in RAM. So reading twice doesn't hurt that much, and maintaining a semblance of consistency in the decompressor api makes it a reasonable tradeoff. 

So we're doing (2)